contracts/src/KeyGen.sol

```diff
@@ -27,300 +29,339 @@ interface IBabyJubjub {
-        require(st.round1[idx].commCoeffs == 0, "Already submitted");
+        if (!_isEmpty(st.round1[partyId].commShare)) revert AlreadySubmitted();
+        if (st.round1[partyId].commCoeffs != 0) revert AlreadySubmitted();
 
         // Add BabyJubJub Elements together and keep running total
-        uint256 pointX = data.commShare.pointX;
-        uint256 pointY = data.commShare.pointY;
-        _addToAggregate(st, pointX, pointY);
+        _addToAggregate(st, data.commShare.x, data.commShare.y);
+        st.round1[partyId] = data;
+        _tryEmitRound2Event(rpId, st);
+    }
 
-        st.round1[idx] = data;
+    function addRound2Contribution(uint128 rpId, Types.Round2Contribution calldata data) external isReady {
+        // check that the contribution is complete
+        if (data.ciphers.length != numPeers) revert BadContribution();  
+        // check that we started the key-gen for this rp-id
+        Types.RpNullifierGenState storage st = runningKeyGens[rpId];
+        if (!st.exists) revert UnknownId(rpId);
+        // check that we are actually in round2
+        if (!st.round2EventEmitted || st.round3EventEmitted) revert WrongRound();
+        // return the partyId if sender is really a participant
+        uint256 partyId = _internParticipantCheck();
+        // check that this peer did not submit anything for this round
+        if (st.round2Done[partyId]) revert AlreadySubmitted();
+
+        // everything looks good - push the ciphertexts
+        for (uint i = 0; i < numPeers; ++i) {
+            st.round2[i][partyId] = data.ciphers[i];
+        }
+        // set the contribution to done
+        st.round2Done[partyId] = true;
+
+        // last step verify the proof and potentially revert if proof fails
+
+        // build the public input:
+        // 1) PublicKey from sender (Affine Point Babyjubjub)
+        // 2) Commitment to share (Affine Point Babyjubjub)
+        // 3) Commitment to coeffs (Basefield Babyjubjub)
+        // 4) Ciphertexts for peers (in this case 3 Basefield BabyJubJub)
+        // 5) Commitments to plaintexts (in this case 3 Affine Points BabyJubJub)
+        // 6) Degree (Basefield BabyJubJub)
+        // 7) Public Keys from peers (in this case 3 Affine Points BabyJubJub)
+        // 8) Nonces (in this case 3 Basefield BabyJubJub)
+        // verifier.verifyProof();
+        uint256[PUBLIC_INPUT_LENGTH_KEYGEN_13] memory publicInputs;
+    
+        publicInputs[0] = peerPublicKeys[partyId].x;
+        publicInputs[1] = peerPublicKeys[partyId].y;
+        publicInputs[2] = st.round1[partyId].commShare.x;
+        publicInputs[3] = st.round1[partyId].commShare.y;
+        publicInputs[4] = st.round1[partyId].commCoeffs;
+        publicInputs[14] = threshold - 1;
+        // peer keys
+        for (uint i = 0; i < numPeers; ++i) {
+            publicInputs[5 + i] = data.ciphers[i].cipher;
+            publicInputs[5 + numPeers + (i * 2) + 0] = data.ciphers[i].commitment.x;
```

Comment:
can we quickly rename the magic numbers 5 15 etc to be constants or just something to make it just a tad more clear.

When we finish ill probably put this logic in a dedicated builder function so incase we need to change this in the future its easily doable and generalizable but for now totally understand the time crunch no worries at all.